src/BuiltInTools/dotnet-watch/HotReload/ScopedCssFileHandler.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Watcher.Internal;
+
+namespace Microsoft.DotNet.Watcher.Tools
+{
+    internal sealed class ScopedCssFileHandler
+    {
+        private static readonly string _muxerPath = new Muxer().MuxerPath;
+        private readonly ProcessRunner _processRunner;
+        private readonly Extensions.Tools.Internal.IReporter _reporter;
+
+        public ScopedCssFileHandler(ProcessRunner processRunner, Extensions.Tools.Internal.IReporter reporter)
+        {
+            _processRunner = processRunner;
+            _reporter = reporter;
+        }
+
+        public async ValueTask<bool> TryHandleFileChange(DotNetWatchContext context, FileItem file, CancellationToken cancellationToken)
+        {
+            if (!file.FilePath.EndsWith(".razor.css", StringComparison.Ordinal) &&
+                !file.FilePath.EndsWith(".cshtml.css", StringComparison.Ordinal))
+            {
+                return default;
+            }
+
+            _reporter.Verbose($"Handling file change event for scoped css file {file.FilePath}.");
+            if (!await RebuildScopedCss(file.ProjectPath, cancellationToken))
+            {
+                return false;
+            }
+            await HandleBrowserRefresh(context.BrowserRefreshServer, file, cancellationToken);
+            return true;
+        }
+
+        private async ValueTask<bool> RebuildScopedCss(string projectPath, CancellationToken cancellationToken)
+        {
+            var build = new ProcessSpec
+            {
+                Executable = _muxerPath,
+                Arguments = new[] { "msbuild", "/nologo", "/t:_PrepareForScopedCss", projectPath, }
+            };
+
+            var result = await _processRunner.RunAsync(build, cancellationToken);
+            return result == 0;
+        }
+
+        private static async Task HandleBrowserRefresh(BrowserRefreshServer browserRefreshServer, FileItem fileItem, CancellationToken cancellationToken)
+        {
+            // We'd like an accurate scoped css path, but this needs a lot of work to wire-up now.
```

Comment:
Does moving the CSS generation to source generators make things any easier?

I'm not super familiar with the EnC/source generators side of the house but wonder if that helps here.

Reply:
It would remove the need to run msbuild and make it faster. But we would still need to know what the path of the css file would be in markup. It's a bit tricky because it's relative to the wwwroot of the project being served. For instance, in a hosted app, the file turns out to be `MyApp.Client.styles.css` even though the project being run is `MyApp.Hosted`. But if the bundle were present in the Shared folder, it's path would be `_content/MyApp.Shared/MyApp.Shared.styles.css`. Lots of oddities to work thru, so taking the easy path and updating everything makes it cheap.